### ISSUE TITLE
dataclasses - mask issue in 2011 simulation (Trac #904)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#898]
In the file /data/sim/IceCube/2011/filtered/level2/CORSIKA-in-ice/10649/00000-00999/Level2_IC86.2011_corsika.010649.000561.i3.bz2, EventID 2142, there is a mask error.

```text
In [3]: for fr in dataio.I3File('/data/sim/IceCube/2011/filtered/level2/CORSIKA-in-ice/10649/00000-00999/Level2_IC86.2011_corsika.010649.000561.i3.bz2'):
    if 'MaskedOfflinePulses' in fr:
        fr['MaskedOfflinePulses'].apply(fr)      
   ...:         
FATAL (I3RecoPulseSeriesMapMask): The mask for OM(56,46) has 1 entries, but source pulse vector has 10 entries! (I3MapOMKeyMask.cxx:480 in boost::shared_ptr<const I3Map<OMKey, std::vector<I3RecoPulse, std::allocator<I3RecoPulse> > > > I3RecoPulseSeriesMapMask::Apply(const I3Frame&) const)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-3-85e79e1bedde> in <module>()
      1 for fr in dataio.I3File('/data/sim/IceCube/2011/filtered/level2/CORSIKA-in-ice/10649/00000-00999/Level2_IC86.2011_corsika.010649.000561.i3.bz2'):
      2     if 'MaskedOfflinePulses' in fr:
----> 3         fr['MaskedOfflinePulses'].apply(fr)
      4 

RuntimeError: The mask for OM(56,46) has 1 entries, but source pulse vector has 10 entries! (in boost::shared_ptr<const I3Map<OMKey, std::vector<I3RecoPulse, std::allocator<I3RecoPulse> > > > I3RecoPulseSeriesMapMask::Apply(const I3Frame&) const)
```

The issue is that the I3SuperDST RecoPulseSeriesMap on OMKey(56,46,0) has more entries than the OfflinePulses mask has entries.  This is an issue in creation of the mask itself.

The software in use is icerec IC2011-L2_V12-08-00_IceSim4compat_V4.

This changeset might be why this doesn't occur anymore: [96060/IceCube]

Suggestions:
* Test recreating the pulse mask with and without that changeset.
* Decide what to do when reading bad masks in the future. Options:
  * remove extra pulses
  * keep extra pulses
  * continue throwing an error, but make it clearer

Migrated from https://code.icecube.wisc.edu/ticket/904
```json
{
    "status": "closed", 
    "changetime": "2015-08-14T08:07:52", 
    "description": "In the file /data/sim/IceCube/2011/filtered/level2/CORSIKA-in-ice/10649/00000-00999/Level2_IC86.2011_corsika.010649.000561.i3.bz2, EventID 2142, there is a mask error.\n\n{{{\nIn [3]: for fr in dataio.I3File('/data/sim/IceCube/2011/filtered/level2/CORSIKA-in-ice/10649/00000-00999/Level2_IC86.2011_corsika.010649.000561.i3.bz2'):\n    if 'MaskedOfflinePulses' in fr:\n        fr['MaskedOfflinePulses'].apply(fr)      \n   ...:         \nFATAL (I3RecoPulseSeriesMapMask): The mask for OM(56,46) has 1 entries, but source pulse vector has 10 entries! (I3MapOMKeyMask.cxx:480 in boost::shared_ptr<const I3Map<OMKey, std::vector<I3RecoPulse, std::allocator<I3RecoPulse> > > > I3RecoPulseSeriesMapMask::Apply(const I3Frame&) const)\n---------------------------------------------------------------------------\nRuntimeError                              Traceback (most recent call last)\n<ipython-input-3-85e79e1bedde> in <module>()\n      1 for fr in dataio.I3File('/data/sim/IceCube/2011/filtered/level2/CORSIKA-in-ice/10649/00000-00999/Level2_IC86.2011_corsika.010649.000561.i3.bz2'):\n      2     if 'MaskedOfflinePulses' in fr:\n----> 3         fr['MaskedOfflinePulses'].apply(fr)\n      4 \n\nRuntimeError: The mask for OM(56,46) has 1 entries, but source pulse vector has 10 entries! (in boost::shared_ptr<const I3Map<OMKey, std::vector<I3RecoPulse, std::allocator<I3RecoPulse> > > > I3RecoPulseSeriesMapMask::Apply(const I3Frame&) const)\n}}}\n\nThe issue is that the I3SuperDST RecoPulseSeriesMap on OMKey(56,46,0) has more entries than the OfflinePulses mask has entries.  This is an issue in creation of the mask itself.\n\nThe software in use is icerec IC2011-L2_V12-08-00_IceSim4compat_V4.\n\nThis changeset might be why this doesn't occur anymore: [96060/IceCube]\n\nSuggestions:\n* Test recreating the pulse mask with and without that changeset.\n* Decide what to do when reading bad masks in the future. Options:\n  * remove extra pulses\n  * keep extra pulses\n  * continue throwing an error, but make it clearer", 
    "reporter": "david.schultz", 
    "cc": "jvansanten", 
    "resolution": "fixed", 
    "_ts": "1439539672935233", 
    "component": "combo core", 
    "summary": "dataclasses - mask issue in 2011 simulation", 
    "priority": "blocker", 
    "keywords": "dataclasses", 
    "time": "2015-03-24T19:54:43", 
    "milestone": "", 
    "owner": "jvansanten", 
    "type": "defect"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20150327T15:20:18: jacobi commented:

> I have a similar problem with the IC2012 L2 meta-project when baseprocessing simulated monopoles.
> 
> 
> 
> ```text
> FATAL (I3RecoPulseSeriesMapMask): The mask for OM(1,9) has 2 entries, but source pulse vector has 3 entries!
> ```
> 
> 
> The problematic event, along with a minimal example, which is a stripped down version of [http://code.icecube.wisc.edu/projects/icecube/browser/IceCube/projects/jeb-filter-2012/trunk/python/baseproc.py baseproc.py] can be found here:
> 
> `/data/user/jacobi/mask_fails`
> 
> The mask `InIceDSTPulses` which acts on `I3SuperDST` is the one failing.
> 
> 
> The example works even on current IceRec (V04-10-00), so r96060/IceCube does not fix mask creations completely. 
>

**Comment by icecube-trac** [comment]
Trac update at 20150327T15:23:07: 
* jacobi changed attachment from "" to "IC86_SLOP_sim-mp_run-000004_n-500_beta-0.001000_lambda-0.100m_event_59_triggered.i3.bz2"
* jacobi commented:

> Event which fails baseprocessing

**Comment by icecube-trac** [comment]
Trac update at 20150327T15:23:56: 
* jacobi commented:

> stripped down version of baseprocesing.py to quickly reproduce the problem


* jacobi changed attachment from "" to "create_bad_mask.py"

**Comment by icecube-trac** [comment]
Trac update at 20150408T14:18:30: olivas changed milestone from "" to "offline V15-03-00"

**Comment by icecube-trac** [comment]
Trac update at 20150409T18:23:10: olivas changed attachment from "" to "mask_test.py"

**Comment by icecube-trac** [comment]
Trac update at 20150409T18:23:22: olivas commented:

> I get a different error trying to read old data with the trunk of IceRec.
> 
> python: /home/olivas/icecube/icerec/trunk/src/dataclasses/private/dataclasses/payload/I3SuperDST.cxx:1149: void I3SuperDST::load_v1(Archive&) [with Archive = boost::archive::portable_binary_iarchive]: Assertion `width_it != width_end' failed.
> Aborted (core dumped)

**Comment by icecube-trac** [comment]
Trac update at 20150414T20:46:12: olivas commented:

> OK...I'm calling this progress.
> 
> With a Release build I can reproduce the original error, because those asserts are compiled out.  But the asserts are real and seem to possibly point in the direction of the actual error, which begs the question...should those be bald 'assert's or 'i3_assert's (i.e. log_fatal wrappers).
> 
> In any case, it's nice to be able to reproduce the error and I'm starting to think these problems are connected.

**Comment by icecube-trac** [comment]
Trac update at 20150414T22:50:30: olivas commented:

> Reverting r96060 doesn't make the problem go away.

**Comment by icecube-trac** [comment]
Trac update at 20150414T23:14:47: olivas commented:

> Rolling back I3SuperDST to r90906 doesn't help either.  I would have been surprised if it had, but it was worth a shot.  That roll-back didn't compile without a small tweak in I3SuperDST::Unpack to account for the fact that I3RecoPulse time changed from a float to a double almost two years ago.
> 
> Index: ../src/dataclasses/private/dataclasses/payload/I3SuperDST.cxx
> ===================================================================
> --- ../src/dataclasses/private/dataclasses/payload/I3SuperDST.cxx	(revision 90906)
> +++ ../src/dataclasses/private/dataclasses/payload/I3SuperDST.cxx	(working copy)
> @@ -332,7 +332,7 @@
>  		/* Ensure that pulses do not overlap. */
>  		for ( ; next < target.second.end(); current++, next++)
>  			current->SetWidth(std::min(current->GetWidth(),
> -			    next->GetTime()-current->GetTime()));
> +                           float(next->GetTime()-current->GetTime())));
>  	}
> 
> 
> Pretty harmless, right?

**Comment by icecube-trac** [comment]
Trac update at 20150416T03:44:31: olivas commented:

> This bug is reproducible at least as far back as V14-02-00, so we've included it in quite a few releases already. Grrr.
> 
> Checking the V13 and V12 releases next.

**Comment by icecube-trac** [comment]
Trac update at 20150506T21:05:16: 
* olivas changed status from "new" to "assigned"
* olivas changed owner from "olivas" to "jvansanten"

**Comment by icecube-trac** [comment]
Trac update at 20150509T17:54:31: 
* jvansanten commented:

> Something has definitely gone wrong with the encoding, but without the original inputs it doesn't seem to be possible to reconstruct what happened. In any case, the assertion is tripped when the iterator over charge stamps gets ahead of the iteration over DOM headers by one unit, and runs off the end of the charge stamps and width markers at the entry for the last DOM to fire in the event. A few observations for the books:
> - The same bug affects 12 of the first 1000 files in dataset 10649. This is a small number, but perhaps not quite small enough to ignore. These are:
> 
>  
> ```text
> /data/sim/IceCube/2011/filtered/level2/CORSIKA-in-ice/10649/00000-00999/Level2_IC86.2011_corsika.010649.000168.i3.bz2
> /data/sim/IceCube/2011/filtered/level2/CORSIKA-in-ice/10649/00000-00999/Level2_IC86.2011_corsika.010649.000188.i3.bz2
> /data/sim/IceCube/2011/filtered/level2/CORSIKA-in-ice/10649/00000-00999/Level2_IC86.2011_corsika.010649.000200.i3.bz2
> /data/sim/IceCube/2011/filtered/level2/CORSIKA-in-ice/10649/00000-00999/Level2_IC86.2011_corsika.010649.000365.i3.bz2
> /data/sim/IceCube/2011/filtered/level2/CORSIKA-in-ice/10649/00000-00999/Level2_IC86.2011_corsika.010649.000480.i3.bz2
> /data/sim/IceCube/2011/filtered/level2/CORSIKA-in-ice/10649/00000-00999/Level2_IC86.2011_corsika.010649.000509.i3.bz2
> /data/sim/IceCube/2011/filtered/level2/CORSIKA-in-ice/10649/00000-00999/Level2_IC86.2011_corsika.010649.000551.i3.bz2
> /data/sim/IceCube/2011/filtered/level2/CORSIKA-in-ice/10649/00000-00999/Level2_IC86.2011_corsika.010649.000561.i3.bz2
> /data/sim/IceCube/2011/filtered/level2/CORSIKA-in-ice/10649/00000-00999/Level2_IC86.2011_corsika.010649.000649.i3.bz2
> /data/sim/IceCube/2011/filtered/level2/CORSIKA-in-ice/10649/00000-00999/Level2_IC86.2011_corsika.010649.000686.i3.bz2
> /data/sim/IceCube/2011/filtered/level2/CORSIKA-in-ice/10649/00000-00999/Level2_IC86.2011_corsika.010649.000717.i3.bz2
> /data/sim/IceCube/2011/filtered/level2/CORSIKA-in-ice/10649/00000-00999/Level2_IC86.2011_corsika.010649.000974.i3.bz2
> ```
> - It seems to be exposed by a trigger-shifter bug: the events that fail have global triggers that start at t>=60 microseconds rather than 0 (this can't happen in data). The global trigger times of the events that fail are:
> 
> 
> ```text
> GLOBAL MERGED 0 configID not set FIRED  length = 21.1546 mus time = 61.0477 mus
> GLOBAL MERGED 0 configID not set FIRED  length = 23.5686 mus time = 125.896 mus
> GLOBAL MERGED 0 configID not set FIRED  length = 21.9221 mus time = 61.1391 mus
> GLOBAL MERGED 0 configID not set FIRED  length = 20 mus time = 60.9864 mus
> GLOBAL MERGED 0 configID not set FIRED  length = 21.6614 mus time = 60.7384 mus
> GLOBAL MERGED 0 configID not set FIRED  length = 22.184 mus time = 60.5235 mus
> GLOBAL MERGED 0 configID not set FIRED  length = 21.8455 mus time = 60.9481 mus
> GLOBAL MERGED 0 configID not set FIRED  length = 20 mus time = 60.4713 mus
> GLOBAL MERGED 0 configID not set FIRED  length = 21.2454 mus time = 69.4716 mus
> GLOBAL MERGED 0 configID not set FIRED  length = 20 mus time = 60.8808 mus
> GLOBAL MERGED 0 configID not set FIRED  length = 20.1449 mus time = 60.7866 mus
> GLOBAL MERGED 0 configID not set FIRED  length = 21.2498 mus time = 61.0134 mus
> ```
> - The exact pulse configuration that trips the assert in that particular file can't be reproduced, because photons are distributed to OpenCL worker threads (each of which uses a different random number sequence) in unspecified order. Bummer.
> - Attempts to artificially trigger the bug by creating events with large time gaps between pulses have been unsuccessful so far. I've tried generating fake "noise" events with ~millisecond gaps between pulses and stretching the time scales of existing events by factors of 100k. The large gap at the beginning of the offending events may be either a red herring or merely a symptom of a different, more complicated problem.


* jvansanten changed _comment0 from:

> Something has definitely gone wrong with the encoding, but without the original inputs it doesn't seem to be possible to reconstruct what happened. In any case, the assertion is tripped when the iterator over charge stamps gets ahead of the iteration over DOM headers by one unit, and runs off the end of the charge stamps and width markers at the entry for the last DOM to fire in the event. A few observations for the books:
> - The same bug affects 14 of the first 1000 files in dataset 10649. This is a small number, but perhaps not quite small enough to ignore.
> - It seems to be exposed by a trigger-shifter bug: the events that fail have global triggers that start at t=~105 microseconds rather than 0 (this can't happen in data)
> - The exact pulse configuration that trips the assert in that particular file can't be reproduced, because photons are distributed to OpenCL worker threads (each of which uses a different random number sequence) in unspecified order. Bummer.
> - Attempts to artificially trigger the bug by creating events with large time gaps between pulses have been unsuccessful so far. I've tried generating fake "noise" events with ~millisecond gaps between pulses and stretching the time scales of existing events by factors of 100k. The large gap at the beginning of the offending events may be either a red herring or merely a symptom of a different, more complicated problem.

to:

> 1432134224979942

**Comment by icecube-trac** [comment]
Trac update at 20150624T11:42:27: jacobi commented:

> Recently I'm running quite often into this bug when working with FixedRateTrigger data. Around FRT 50 runs from 2012 fail due to mask errors.
> 
> In a debug build the following assert gets triggered
> 
> 
> 
> ```text
> python: /afs/ifh.de/user/j/jacobi/software/IceRec_V04-10-00/src/dataclasses/private/dataclasses/payload/I3SuperDST.cxx:1161: void I3SuperDST::load_v1(Archive&) [with Archive = boost::archive::portable_binary_iarchive]: Assertion `stamp_it->stamp.hlc_bit == hlc' failed.
> /afs/ifh.de/user/j/jacobi/software/IceRec_V04-10-00/build_debug/env-shell.sh: line 155: 18793 Aborted
> ```
>

**Comment by icecube-trac** [comment]
Trac update at 20150625T12:27:04: 
* icecube changed attachment from "" to "Level1_IC86.2012_data_Run00120159_FRT_SLOP_event_688.i3.bz2"
* icecube commented:

> SLOP triggered FRT data from run 120159. This file still contains the launches, which should help debugging. When baseprocessing this event a bad mask is created (InIcePulses). Use the minimal example script from above to reproduce the problem.

**Comment by icecube-trac** [comment]
Trac update at 20150707T21:09:49: 
* olivas changed milestone from "offline V15-05-00" to "offline V15-07-00"
* olivas changed priority from "major" to "blocker"

**Comment by icecube-trac** [comment]
Trac update at 20150810T20:31:17: 
* olivas commented:

> Milestone offline V15-07-00 deleted


* olivas changed milestone from "offline V15-07-00" to ""

**Comment by icecube-trac** [comment]
Trac update at 20150813T15:12:14: jvansanten commented:

> Thanks to Emanuel's test case, I think I've tracked this down. There's a bug in the variable-width encoding of time delays between the first pulse times on each DOM. The time delay since the last pulse is encoded as an 11-bit number. If the actual time delay is > 2047, then the remainder is added in 16-bit chunks. If the decoder encounters a first time of 2047, it will interpret the following blocks as the remainder until one is encountered with a value of < 65535. 
> 
> If the time delay in ns rounded as an integer is exactly 67582, however, the encoder failed to emit a 0 remainder block to mark the end of the run, so the decoder interprets the block corresponding to the next pulse as a remainder.
> 
> The bad news is that SuperDST payloads encoded this way are in fact irredeemably corrupted. The good news is that I think this really does only affect simulation. I'm still not sure how such large time delays ended up in Emanuel's scrambles, however.

**Comment by icecube-trac** [comment]
Trac update at 20150813T16:02:00: david.schultz commented:

> Some people are worried about detecting this in current files.  Can we develop a test script for this so people can check their files?  Would also be good to add a "bad" event to the test-data so we can add it to the regular test suite as well.

**Comment by icecube-trac** [comment]
Trac update at 20150814T08:07:52: 
* jvansanten changed resolution from "" to "fixed"
* jvansanten changed status from "assigned" to "closed"
* jvansanten commented:

> In changeset 136207:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="136207"
> Fix an encoding bug that occured whenever the time delay between the first
> pulses on successive DOMs was between 2047+N*65535 and 2047+N*65535+1
> ns, where N is a natural number. This was observed in buggy simulation,
> but can't happen in real data. A similar vulnerability existed in the
> charge overflow encoding, but is much harder to trigger.
> 
> Along the way, promote all assertions in I3SuperDST to i3_assert()
> so that this never happens again.
> 
> Closes #898, though as noted in the ticket, there is no way to recover
> events affected by the bug.
> ```

**Comment by icecube-trac** [comment]
Trac update at 20150814T08:37:29: 
* jvansanten changed attachment from "" to "harvest_corrupted_superdst.py"
* jvansanten commented:

> A script to check for corruption. This should be used either with the noted metaproject or with dataclasses trunk, where asserts() have been promoted to i3_assert()
